_posts/2023-17-07-gams-blog.md

```diff
@@ -0,0 +1,373 @@
+---
+layout: post
+title: "JuMP, GAMS, and the IJKLM model"
+date: 2023-17-07
+categories: [tutorials]
+---
+# JuMP, GAMS, and the IJKLM model
+
+A [recent blog post](https://www.gams.com/blog/2023/07/performance-in-optimization-models-a-comparative-analysis-of-gams-pyomo-gurobipy-and-jump)
+by GAMS demonstrated a significant performance difference between JuMP and
+GAMS on a model they call IJKLM. The purpose of this blog post is to explain
+the difference and present a different formulation with much better
+performance.
+
+Read the GAMS post for background on the model, as well as its mathematical
+formulation.
+
+## The input data
+
+Input data for the model is provided in three JSON files, labeled IJK, JKL,
+and KLM.
+
+````julia
+import JSON
+
+const DATA_DIR = joinpath(@__DIR__, "gams");
+
+struct OriginalData
+    I::Vector{String}
+    IJK::Vector{Vector{Any}}
+    JKL::Vector{Vector{Any}}
+    KLM::Vector{Vector{Any}}
+
+    function OriginalData(n::Int)
+        return new(
+            ["i$i" for i in 1:n],
+            JSON.parsefile(joinpath(DATA_DIR, "data_IJK_$n.json")),
+            JSON.parsefile(joinpath(DATA_DIR, "data_JKL.json")),
+            JSON.parsefile(joinpath(DATA_DIR, "data_KLM.json")),
+        )
+    end
+end
+````
+
+The GAMS blog post generated a range of different sizes, but we'll just use
+the `n = 2200` case. You can use the code in GAMS's [GitHub repository](https://github.com/justine18/performance_experiment)
```

Comment:
Added an asymptotic post out to N=3x10^3. The blogpost goes out to N=4x10^5, but I don't have time to run that on my laptop. it's a bit of a pain, because you need to run the original code to build the JSON files.

Reply:
I don't know if we want to get into the business of running the gurobipy and pyomo models, since they could also be updated.

Reply:
Ok, we don't need to rerun gurobipy and pyomo. I think it's sufficient to note that the new model is asymptotically better, which would indeed affect the qualitative conclusions made by the GAMS post.